torchmultimodal/models/mdetr.py

```diff
@@ -0,0 +1,676 @@
+                query_pos=query_pos,
+            )
+            if self.return_intermediate:
+                if self.norm is not None:
+                    intermediate.append(self.norm(output))
+                else:
+                    intermediate.append(output)
+
+        if self.return_intermediate:
+            return torch.stack(intermediate)
+
+        if self.norm is not None:
+            return self.norm(output)
+
+        return output
+
+
+class TransformerEncoderLayer(nn.Module):
+    """
+    A single layer from a transformer encoder.
+
+    Args:   d_model (int): Number of features in the input.
+            nhead (int): Number of heads in multi-head attention.
+            dim_feedforward (int): Dimension of feedforward network. Default: 2048
+            dropout (float): Dropout value. Default: 0.1.
+            activation (Callable[..., Tensor]): The activation function of the
+                intermediate layer. Default: relu
+            normalize_before (bool): Whether to do PreNorm. Default: False
+
+    Inputs: src (Tensor): The sequence to the encoder layer.
+            src_mask (Optional[Tensor]) The mask for the src sequence. Default: None
+            src_key_padding_mask (Optional[Tensor]): The mask for the src keys per batch.
+                Default: None
+            pos (Optional[Tensor]): Positional embeddings applied to Q and K
+                self-attention matrices. Default: None
+    """
+
+    def __init__(
+        self,
+        d_model: int,
+        nhead: int,
+        dim_feedforward: int = 2048,
+        dropout: float = 0.1,
+        activation: Callable[..., Tensor] = nn.functional.relu,
+        normalize_before: bool = False,
+    ):
+        super().__init__()
+        self.self_attn = nn.MultiheadAttention(d_model, nhead, dropout=dropout)
+        # Implementation of Feedforward model
+        self.linear1 = nn.Linear(d_model, dim_feedforward)
+        self.dropout = nn.Dropout(dropout)
+        self.linear2 = nn.Linear(dim_feedforward, d_model)
+
+        self.norm1 = nn.LayerNorm(d_model)
+        self.norm2 = nn.LayerNorm(d_model)
+        self.dropout1 = nn.Dropout(dropout)
+        self.dropout2 = nn.Dropout(dropout)
+
+        self.activation = activation
+        self.normalize_before = normalize_before
+
+    def with_pos_embed(self, tensor: Tensor, pos: Optional[Tensor]) -> Tensor:
+        return tensor if pos is None else tensor + pos
+
+    def forward_post(
+        self,
+        src: Tensor,
+        src_mask: Optional[Tensor] = None,
+        src_key_padding_mask: Optional[Tensor] = None,
+        pos: Optional[Tensor] = None,
+    ) -> Tensor:
+        q = k = self.with_pos_embed(src, pos)
+        src2 = self.self_attn(
+            q, k, value=src, attn_mask=src_mask, key_padding_mask=src_key_padding_mask
+        )[0]
+        src = src + self.dropout1(src2)
+        src = self.norm1(src)
+        src2 = self.linear2(self.dropout(self.activation(self.linear1(src))))
```

Comment:
turn this into nn sequential or use mlp to make it more readable? also prevent overwriting src and rename src2